ambry-frontend/src/main/java/com/github/ambry/frontend/NamedBlobDeleteHandler.java

```diff
@@ -0,0 +1,105 @@
+    this.metrics = metrics;
+    this.deleteBlobHandler = deleteBlobHandler;
+  }
+
+  /**
+   * Asynchronously handle named blob delete requests.
+   * @param restRequest the {@link RestRequest} that contains the request parameters and body.
+   * @param restResponseChannel the {@link RestResponseChannel} where headers should be set.
+   * @param callback the {@link Callback} to invoke when the response is ready (or if there is an exception).
+   * @throws RestServiceException
+   */
+  void handle(RestRequest restRequest, RestResponseChannel restResponseChannel, Callback<Void> callback)
+      throws RestServiceException {
+    accountAndContainerInjector.injectAccountAndContainerForNamedBlob(restRequest, metrics.deleteBlobMetricsGroup);
+    RequestPath requestPath = getRequestPath(restRequest);
+    // Get blob id and reconstruct RequestPath
+    // this is the three-part named blob id, including account name, container name and the custom blob name.
+    String namedBlobId = requestPath.getOperationOrBlobId(false);
+    NamedBlobPath namedBlobPath = NamedBlobPath.parse(namedBlobId, Collections.emptyMap());
+    // This is a three steps operation
+    // 1. Delete database row.
+    // 2. Delete regular blob id from Router
+    // Ie we failed at step 1, or step 1 returns NotFound, then  we just return the error back to client, indicating failure
+    // If we failed at step 2, client can resend the request to ambry so we can resume the operation. If regular blob is already
+    //    deleted, Router would do a nop and return success
+    // The reason why we don't do this in a get-RouterDelete-delete is because in between get and delete, if another client sends
+    // delete and put, then we will have a totally different blob id under this blob name. And we shouldn't delete this new blob id
+    // from database.
+    namedBlobDb.delete(namedBlobPath.getAccountName(), namedBlobPath.getContainerName(), namedBlobPath.getBlobName())
+        .thenApply(DeleteResult::getBlobId)
+        .whenComplete((blobId, exception) -> {
+          if (exception != null) {
+            Exception dbException = Utils.extractFutureExceptionCause(exception);
+            callback.onCompletion(null, dbException);
+            return;
+          }
+          // What if this blobName is already deleted?
+          // We still want to issue DELETE to router, since router might still have a valid blob because we delete blobName
+          // from database before deleting blob from router. We might fail after we delete from database and the client retries
+          // delete. Even if client doesn't retry delete, we still have a background job to delete the orphan blob from router.
+          RequestPath newRequestPath =
+              new RequestPath(requestPath.getPrefix(), requestPath.getClusterName(), requestPath.getPathAfterPrefixes(),
+                  "/ " + blobId, requestPath.getSubResource(), requestPath.getBlobSegmentIdx());
```

Comment:
QQ: there's a space after slash, would that be a problem?

Reply:
thank you for pointing this out, now the question is, why the tests worked??

Reply:
I got it now, base64 decoder would ignore the leading space ....